File: Domains/Base/Workspace/CSharp/Domain/Base/Data/Instanceof.cs
//------------------------------------------------------------------------------------------------- 
// <copyright file="Instanceof.cs" company="Allors bvba">
// Copyright 2002-2017 Allors bvba.
// 
// Dual Licensed under
//   a) the Lesser General Public Licence v3 (LGPL)
//   b) the Allors License
// 
// The LGPL License is included in the file lgpl.txt.
// The Allors License is an addendum to your contract.
// 
// Allors Platform is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// For more information visit http://www.allors.com/legal
// </copyright>
//-------------------------------------------------------------------------------------------------

namespace Allors.Workspace.Data
{
    using System.Collections.Generic;

    using Allors.Meta;
    using Allors.Workspace.Data.Protocol;

    public class Instanceof : IPropertyPredicate
    {
        public string Parameter { get; set; }

        public Instanceof(IComposite objectType = null)
        {
            this.ObjectType = objectType;
        }

        public IComposite ObjectType { get; set; }

        public IPropertyType PropertyType { get; set; }

        public Predicate Save()
        {
            return new Predicate
            {
                Kind = PredicateKind.Instanceof,
                ObjectType = this.ObjectType?.Id,
                PropertyType = this.PropertyType?.Id
            };
        }

        bool IPredicate.ShouldTreeShake(IReadOnlyDictionary<string, object> arguments)
        {
            return ((IPredicate)this).HasMissingArguments(arguments);
        }

        bool IPredicate.HasMissingArguments(IReadOnlyDictionary<string, object> arguments)
        {
            return this.Parameter != null && (arguments == null || !arguments.ContainsKey(this.Parameter));
        }
    }
}

File: Platform/Workspace/Typescript/framework/data/Instanceof.ts
import { ParameterizablePredicate, ParameterizablePredicateArgs } from './ParameterizablePredicate';
import { PropertyType } from '../meta/PropertyType';
import { ObjectType } from '../meta/ObjectType';

export interface InstanceofArgs extends ParameterizablePredicateArgs, Pick<Instanceof, 'propertyType' | 'objectType'> {}

export class Instanceof extends ParameterizablePredicate {
  propertyType: PropertyType;
  instanceObjectType?: ObjectType;

  constructor(propertyType: PropertyType);
  constructor(args: InstanceofArgs);
  constructor(args: InstanceofArgs | PropertyType) {
    super();

    if ((args as PropertyType).objectType) {
      this.propertyType = args as PropertyType;
    } else {
      Object.assign(this, args);
      this.propertyType = (args as InstanceofArgs).propertyType;
    }
  }

  get objectType(): ObjectType {
    return this.propertyType.objectType;
  }

  toJSON(): any {
    return {
      kind: 'Instanceof',
      dependencies: this.dependencies,
      propertytype: this.propertyType.id,
      parameter: this.parameter,
      objecttype: this.instanceObjectType?.id,
    };
  }
}
